packages/utils/src/memo.ts

```diff
@@ -0,0 +1,56 @@
+// tslint:disable:no-unsafe-any
+/**
+ * Memo class used for decycle json objects. Uses WeakSet if available otherwise array.
+ */
+export class Memo {
+  /** Determines if WeakSet is available */
+  private readonly hasWeakSet: boolean;
+  /** Either WeakSet or Array */
+  private readonly inner: any;
+
+  public constructor() {
+    // tslint:disable-next-line
+    this.hasWeakSet = typeof WeakSet === 'function';
+    this.inner = this.hasWeakSet ? new WeakSet() : [];
+  }
+
+  /**
+   * Sets obj to remember.
+   * @param obj Object to remember
+   */
+  public memoize(obj: any): boolean {
+    if (this.hasWeakSet) {
+      if (this.inner.has(obj)) {
+        return true;
+      }
+      this.inner.add(obj);
+      return false;
+    } else {
+      for (const value of this.inner) {
+        if (value === obj) {
+          return true;
+        }
+      }
+      this.inner.push(obj);
+      return false;
+    }
+  }
+
+  /**
+   * Removes object from internal storage.
+   * @param obj Object to forget
+   */
+  public unmemoize(obj: any): void {
+    if (this.hasWeakSet) {
+      this.inner.delete(obj);
+    } else {
+      this.inner.forEach((value: any, idx: number) => {
+        if (value === obj) {
+          this.inner.splice(idx, 1);
+          return false;
```

Comment:
It's impossible to break `forEach` iteration so `return` statements are redundant here. We should use `for` instead which can use `break` call.